File: Quant/MAGMA/Bipartite2General.cpp
// ------------------------------------------------------------------------------------------------
// system
#include <iostream>
#include <cstring>
#include <cmath>

#include "magma-2.3.0/include/magma_v2.h"
// ------------------------------------------------------------------------------------------------
// BipartiteGeneral
#include "Bipartite2General/Cavity.h"
#include "Bipartite2General/Hamiltonian.h"

// #include "BipartiteGeneral/WaveFunction.h"
#include "Bipartite2General/DensityMatrix.cpp"

#include "Bipartite2General/Evolution.cpp"

#include "Bipartite2General/parser.cpp" // config
// ------------------------------------------------------------------------------------------------
// Import
#include "Common/ext.cpp"
// ------------------------------------------------------------------------------------------------

using namespace std;

// ------------------------------------------------------------------------------------------------
int main(int argc, char** argv) {
	// --------------------------------------------------------------------------------------------
	magma_init();

	magma_queue_t queue = NULL;
	magma_int_t dev = 0;

	magma_queue_create(dev, &queue);
	// --------------------------------------------------------------------------------------------
	const char* config_filename = (argc > 1) ? (argv[1]) : ("Bipartite2General/config");

	BG2config config;
	config.read_from_file(config_filename);
	print("Parse config... done\n\n", "yellow");
	// --------------------------------------------------------------------------------------------
	mkdir(config.path());

	string cmd = "cp " + string(config_filename) + " " + config.path() + "/config.py";

	if (system(cmd.c_str())) {
		Assert(false, "Error: system(cmd.c_str())", __FILE__, __LINE__);
	}
	// --------------------------------------------------------------------------------------------
	Cavity cv_1(config.n_1(), config.wc_1(), config.wa_1(), config.g_1());
	Cavity cv_2(config.n_2(), config.wc_2(), config.wa_2(), config.g_2());

	cv_1.print_n();
	cv_1.print_wc();
	cv_1.print_wa();
	cv_1.print_g();

	cv_2.print_n();
	cv_2.print_wc();
	cv_2.print_wa();
	cv_2.print_g();
	// --------------------------------------------------------------------------------------------
	Hamiltonian H(config.capacity(), cv_1, cv_2, config.mu(), &queue);
	print("Hamiltonian... done\n\n", "yellow");

	#if DEBUG
	// H.write_to_file(H_csv, config.precision());

	// H.print_states();
	// H.print_bin_states();
	// H.print_html();
	#endif
	// --------------------------------------------------------------------------------------------
	DensityMatrix ro_0(&H, H.states());

	ro_0.set_ampl(config.init_state(), MAGMA_Z_MAKE(1, 0));
	print("Density matrix... done\n\n", "yellow");

	#if DEBUG
	// ro_0.write_to_file(config.ro0_csv, config.precision());
	#endif
	// --------------------------------------------------------------------------------------------
	run(ro_0, H, config.dt(), config);
	print("Evolution... done\n\n", "yellow");

	print("\nSUCCESS\n");
	// --------------------------------------------------------------------------------------------
	magma_queue_destroy(queue);

	magma_finalize();
	// --------------------------------------------------------------------------------------------
	return 0;
}

File: Quant/Python/Bipartite2General.py
# -------------------------------------------------------------------------------------------------
# system
import sys
# -------------------------------------------------------------------------------------------------
# Bipartite2General
from Bipartite2General.Cavity import *
from Bipartite2General.Hamiltonian import *

from Bipartite2General.WaveFunction import *
from Bipartite2General.DensityMatrix import *

from Bipartite2General.Evolution import *
# -------------------------------------------------------------------------------------------------
# Common
from Common.LoadPackage import *
from Common.Print import *

from Common.ext import mkdir
from Common.PyPlot import PyPlot3D
# -------------------------------------------------------------------------------------------------
config = load_pkg("config", "Bipartite2General/config.py")

mkdir(config.path)
# -------------------------------------------------------------------------------------------------
cavity_1 = Cavity(n=config.n_1, wc=config.wc_1, wa=config.wa_1, g=config.g_1)
cavity_2 = Cavity(n=config.n_2, wc=config.wc_2, wa=config.wa_2, g=config.g_2)

print("Cavity 1:\n", color="yellow")
cavity_1.print_n()
cavity_1.print_wc()
cavity_1.print_wa()
cavity_1.print_g()

print()

print("Cavity 2:\n", color="yellow")
cavity_2.print_n()
cavity_2.print_wc()
cavity_2.print_wa()
cavity_2.print_g()
# -------------------------------------------------------------------------------------------------
H = Hamiltonian(capacity=config.capacity, cavity=[
                cavity_1, cavity_2], mu=config.mu)
# H.matrix.set_header(H.states)
# print("L2:\n", color="yellow")
# H.matrix.print_pd()

if __debug__:
    H.write_to_file(filename=config.H_csv)
    # H.print_html()

    print("Hamiltonian states:\n", color="yellow")
    H.print_states()
    H.print_bin_states()
# -------------------------------------------------------------------------------------------------
w_0 = WaveFunction(states=H.states, init_state=config.init_state)

if __debug__:
    print("wf\n", color="yellow")
    w_0.print()
# -------------------------------------------------------------------------------------------------
ro_0 = DensityMatrix(w_0)

if __debug__:
    ro_0.write_to_file(filename=config.ro_0_csv)
# -------------------------------------------------------------------------------------------------

run(ro_0=ro_0, H=H, dt=config.dt, nt=config.nt, config=config)

# -------------------------------------------------------------------------------------------------
from Common.STR import *

if not __debug__:
    title = "<b>"
    # title += "<br>wc = " + str(wc / GHz) + " GHz"    # title += "<br>wa = "
    # + str(np.array(wa) / GHz) + " GHz"    # title += "<br>g = " +
    # str(np.array(g) / MHz) + " MHz"
    title = "<b>Bipartite oscillations</b>"
    title += "<br>capacity = " + \
        str(config.capacity) + ", n_1 = " + \
        str(config.n_1) + ", n_2 = " + str(config.n_2)
    title += "<br>t = " + T_str(config.T)
    title += "</b>"

    PyPlot3D(
        title=title,
        z_csv=config.path + "/" + "z.csv",
        x_csv=config.path + "/" + "x.csv",
        y_csv=config.path + "/" + "t.csv",
        online=False,
        path=config.path,
        filename="Bipartite2",
        xaxis="states",
        yaxis="time, мкс",
        y_scale=5,
    )
# -------------------------------------------------------------------------------------------------
